source/cl/source/extension/source/intel_unified_shared_memory/intel_unified_shared_memory.cpp

```diff
@@ -418,6 +435,148 @@ cl_int device_allocation_info::allocate(cl_uint alignment) {
 
   return CL_SUCCESS;
 }
+
+shared_allocation_info::shared_allocation_info(const cl_context context,
+                                               const cl_device_id device,
+                                               const size_t size)
+    : allocation_info(context, size),
+      device(device),
+      mux_memory(nullptr),
+      mux_buffer(nullptr) {
+  if (device) {
+    cl::retainInternal(device);
+  }
+};
+
+shared_allocation_info::~shared_allocation_info() {
+  if (mux_buffer) {
+    assert(device);
+    (void)muxDestroyBuffer(device->mux_device, mux_buffer,
+                           device->mux_allocator);
+  }
+
+  if (mux_memory) {
+    assert(device);
+    muxFreeMemory(device->mux_device, mux_memory, device->mux_allocator);
+  }
+
+  // Free the host side allocation
+  if (base_ptr) {
+    cargo::free(base_ptr);
+  }
+
+  if (device) {
+    cl::releaseInternal(device);
+  }
+};
+
+cargo::expected<std::unique_ptr<shared_allocation_info>, cl_int>
+shared_allocation_info::create(cl_context context, cl_device_id device,
+                               const cl_mem_properties_intel *properties,
+                               size_t size, cl_uint alignment) {
+  OCL_CHECK(size == 0, return cargo::make_unexpected(CL_INVALID_BUFFER_SIZE));
+
+  if (device) {
+    OCL_CHECK(size > device->max_mem_alloc_size,
+              return cargo::make_unexpected(CL_INVALID_BUFFER_SIZE));
+    OCL_CHECK(alignment > device->mux_device->info->buffer_alignment,
+              return cargo::make_unexpected(CL_INVALID_VALUE));
+  } else {
+    // If no device is given, all devices that support shared USM must meet
+    // the max_alloc_size and alignment requirements
+    for (auto device : context->devices) {
+      if (deviceSupportsSharedAllocations(device)) {
+        const auto device_align = device->mux_device->info->buffer_alignment;
+        OCL_CHECK(size > device->max_mem_alloc_size,
+                  return cargo::make_unexpected(CL_INVALID_BUFFER_SIZE));
+        OCL_CHECK(alignment > device_align,
```

Comment:
Does this check and the line 482 check mean alignment == 0 won't get to line 499+?